File: languages/go/03-advanced/templates-basics/example/generic-print-and-sum.go
// This helper example focuses on isolating the reusable generic behavior before the example wires it together.

// This extra example extends templates-basics with reusable typed helpers.

package main

import "fmt"

// Keep this helper separate so the main example can focus on the larger idea without extra noise.
type number interface {
	~int | ~float64
}

func printSlice[T any](values []T, label string) {
	fmt.Printf("%s: [", label)
	for index, value := range values {
		if index > 0 {
			fmt.Print(", ")
		}
		fmt.Print(value)
	}
	fmt.Println("]")
}

func sumSlice[T number](values []T) T {
	var total T
	for _, value := range values {
		total += value
	}
	return total
}

func main() {
	numbers := []int{2, 4, 6, 8}
	prices := []float64{1.5, 2.25, 3.75}

	printSlice(numbers, "Numbers")
	printSlice(prices, "Prices")

	fmt.Printf("Sum of numbers: %d\n", sumSlice(numbers))
	fmt.Printf("Sum of prices: %.2f\n", sumSlice(prices))
}

File: languages/cpp/03-advanced/templates-basics/example/generic-print-and-sum.cpp
// This helper example focuses on isolating the reusable generic behavior before the example wires
// it together.

#include <iostream>
#include <string>
#include <vector>

using namespace std;

// Keep this helper separate so the main example can focus on the larger idea without extra noise.
template <typename T> void printVector(const vector<T>& values, const string& label) {
    cout << label << ": [";
    for (size_t i = 0; i < values.size(); ++i) {
        cout << values[i];
        if (i + 1 < values.size()) {
            cout << ", ";
        }
    }
    cout << "]\n";
}

template <typename T> T sumVector(const vector<T>& values) {
    T total = T{};
    for (const T& value : values) {
        total += value;
    }
    return total;
}

int main() {
    vector<int> numbers = {2, 4, 6, 8};
    vector<double> prices = {1.5, 2.25, 3.75};

    printVector(numbers, "Numbers");
    printVector(prices, "Prices");

    cout << "Sum of numbers: " << sumVector(numbers) << '\n';
    cout << "Sum of prices: " << sumVector(prices) << '\n';

    return 0;
}
